### PR TITLE
feat: implement asynchronous scheduling with task pipeline.

### PR DESCRIPTION
**Pull request by oleanderrr**
## Description

Add an opt-in task pipeline for Python Qwen3 and GLM inference on Ascend NPU. It separates input preparation, ordered model execution, and result collection so that preparation for the next batch can overlap execution of the current batch, while each in-flight task retains its own input and result storage.

### Implementation

The implementation has two core components:

- **`TaskExecutionPipeline`** manages task admission, FIFO ordering, backpressure, execution, and slot retirement. Prepare and Consume run on the state executor; a dedicated launch thread submits model work in task order. Prepare, model execution, and result copies use separate NPU streams with event dependencies.
- **`SlotBuffer`** owns a slot's reusable pinned-host/device model inputs, sampling inputs, attention metadata, previous-token mappings, and single-token results, including optional logprobs. Buffers are sized from the existing batch/model limits and allocated before KV-cache budgeting.

The execution flow is:

1. **Prepare:** accept the existing `ForwardInput`, unpack transport data inside the pipeline, validate it, stage inputs, and prepare task-local Python attention metadata. `submit()` returns only after the pipeline owns the input data; the caller's input storage is no longer borrowed.
2. **Launch:** wait for input readiness, resolve the scheduler's negative token placeholders from the previous sampling output, and run the existing model, logits, and sampler paths. Model execution remains ordered on one task stream. A shared previous-token buffer survives retirement of the producing slot.
3. **Consume:** copy results through the result stream, wait for completion, return an independent CPU `ForwardOutput`, and retire the oldest slot before completing its result future. An unconsumed result continues to occupy its slot.

Python attention metadata preparation is separated from activation, so preparing the next batch does not overwrite the backend state used by the current forward. Worker directly passes `ForwardInput`/`ForwardOutput`; WorkerService uses Prepare acknowledgement for overlap and consumes the ready CPU results. In multi-rank execution, the engine collects completion from every worker to retire all slots, while forwarding only driver outputs to the scheduler.

### Configuration and scope

`enable_task_pipeline` defaults to `false`. The existing `enable_schedule_overlap` determines the number of slots:

| `enable_task_pipeline` | `enable_schedule_overlap` | Execution path |
| --- | --- | --- |
| `false` | `false` / `true` | Existing execution path |
| `true` | `false` | Single slot |
| `true` | `true` | Two slots |

For dual-slot execution, add:

```sh
--enable_task_pipeline=true
--enable_schedule_overlap=true
--model_impl=python
```

### Performance comparison

GLM5.3 W8A8/BF16 on 16 Ascend910 logical NPUs, TP16 / EP16 / DP1, Python eager. Each case uses concurrency **1**, **2048 input tokens**, **256 output tokens**, **5 warmup requests**, then **10 measured requests**, without profiling. Inputs are identical pre-encoded token IDs; `temperature=0`, `ignore_eos=true`, and seed 1234. Prefix cache is disabled; chunked prefill is enabled. Both sides use the same batch limits (8192 tokens, 16 sequences) and KV block size (128).

**main + overlap vs. dual slot**

| Client metric, mean of 10 requests | main + overlap | Dual slot | Change |
| --- | ---: | ---: | ---: |
| End-to-end latency (s) | 42.070 | 41.525 | -1.30% |
| TPOT (ms) | 162.091 | 159.952 | -1.32% |
| TTFT (ms) | 737.332 | 736.936 | -0.05% |

**main without overlap vs. single slot**

| Client metric, mean of 10 requests | main, overlap off | Single slot | Change |
| --- | ---: | ---: | ---: |
| End-to-end latency (s) | 42.940 | 42.884 | -0.13% |
| TPOT (ms) | 165.494 | 165.280 | -0.13% |
| TTFT (ms) | 739.276 | 737.351 | -0.26% |

## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Performance improvement
- [ ] Refactor
- [ ] Documentation
- [ ] Test
- [ ] Build or CI

## Pull Request Checklist

Thank you for contributing to xLLM. Before requesting review, please make sure the following items are complete.

### PR Title and Commit Messages

- [x] The PR title and each commit message follow the xLLM commit format: `<type>: <subject>`.

> Allowed types: feat, bugfix, docs, test, refactor, chore, style, revert, perf, model, build, release.
> The subject should use clear English, start with a verb, include at least 4 words, and end with `.`.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` or an equivalent command.
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

### Self Review

- [x] I have self-reviewed the code according to `.agents/skills/code-review/references/custom-code-style.md`, especially code written or assisted by AI.
- [ ] I have rebased this PR onto the latest `main` branch.

### Build and Test Coverage

- [ ] Tests have been added or updated as needed.
- [ ] CUDA: `python setup.py build test` has passed on a CUDA machine.
- [ ] NPU: `python setup.py build test` has passed on an NPU machine.
- [ ] MLU: `python setup.py build test` has passed on an MLU machine.

### CODE REVIEW

File: xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+    return invalid("Task pipeline requires one state thread.");
+  }
+  if (!executor.supports_prepared_attention_metadata()) {
+    return invalid(
+        "LLM task pipeline requires supported Python prepared metadata.");
+  }
+  if (capacity.slot_count == 0 || capacity.slot_count > 2 ||
+      capacity.model.max_sequences == 0 ||
+      capacity.model.max_sequences > std::numeric_limits<int32_t>::max() ||
+      model.device().type() != torch::kPrivateUse1 ||
```
Comment (XuZhang99) [inline xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp]: use `class Platform` to detect hardware:
https://github.com/xLLM-AI/xllm/blob/a48d4350881aff07f78f0c0c283c88b1ee7f35ad/xllm/core/platform/platform.h#L39-L43

File: xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+  return Status(StatusCode::INVALID_ARGUMENT, message);
+}
+
+StreamEventPtr reusable_event() {
+  aclrtEvent event = nullptr;
+  CHECK_EQ(aclrtCreateEventExWithFlag(&event, ACL_EVENT_SYNC), ACL_SUCCESS);
+  return std::make_shared<StreamEvent>(event);
+}
+
+void record(const Stream& stream, const StreamEventPtr& event) {
```
Comment (XuZhang99) [inline xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp]: could move this func to `class Stream`?

File: xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+           ACL_SUCCESS);
+}
+
+bool cpu_int_tensor(const torch::Tensor& tensor, int32_t dimensions) {
+  return tensor.defined() && tensor.device().is_cpu() &&
+         tensor.scalar_type() == torch::kInt32 && tensor.dim() == dimensions &&
+         tensor.is_contiguous();
+}
+
+std::span<const int32_t> int_span(const torch::Tensor& tensor) {
```
Comment (XuZhang99) [inline xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp]: move `int_span` and `cpu_int_tensor`(rename to `is_cpu_int_tensor`) to `tensor_helper.h`.

File: xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+    if (!status.ok()) {
+      return status;
+    }
+    slot->input_ready = reusable_event();
+    slot->output_ready = reusable_event();
+    pipeline->slots_.emplace_back(std::move(slot));
+  }
+  // Initialization is outside serving. Complete allocation-stream writes
+  // before handing storage to either execution thread.
+  Stream initialization(c10_npu::getCurrentNPUStream(model.device().index()));
```
Comment (XuZhang99) [inline xllm/core/runtime/task_pipeline/task_execution_pipeline.cpp]: I think it's better to add `xllm::Device device_` to `TaskExecutionPipeline` and get current stream using `current_stream()`.

File: xllm/core/runtime/CMakeLists.txt [hunk 5]
```diff
@@ -1,6 +1,13 @@
 include(cc_library)
 include(cc_binary)
 
+cc_library(
+  NAME task_pipeline_slot_buffer
+  HDRS task_pipeline/slot_buffer.h
```
Comment (XuZhang99) [inline xllm/core/runtime/CMakeLists.txt]: why not add `slot_buffer.h` and `slot_buffer.cpp` to `runtime` directly?

**Review by XuZhang99 — APPROVED** [review]
